File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/proxy/RtpSenderProxy.kt
package com.instrumentisto.medea_flutter_webrtc.proxy

import com.instrumentisto.medea_flutter_webrtc.exception.ReplaceSenderTrackException
import org.webrtc.RtpSender

/**
 * Wrapper around an [RtpSender].
 *
 * @param sender Actual underlying [RtpSender].
 */
class RtpSenderProxy(sender: RtpSender) : Proxy<RtpSender>(sender) {

  /** [MediaStreamTrackProxy] of this [RtpReceiverProxy]. */
  private var track: MediaStreamTrackProxy? = null

  /** TODO */
  private var disposed: Boolean = false

  init {
    syncMediaStreamTrack()
    addOnSyncListener { syncMediaStreamTrack() }
  }

  /** TODO */
  fun setDisposed() {
    disposed = true
  }

  /**
   * Replaces [MediaStreamTrackProxy] of the underlying [RtpSender] with the provided one.
   *
   * @param t [MediaStreamTrackProxy] which will be set to the underlying [RtpSender].
   */
  fun replaceTrack(t: MediaStreamTrackProxy?) {
    if (!disposed) {
      track = t
      val isSuccessful = obj.setTrack(t?.obj, false)
      if (!isSuccessful) {
        throw ReplaceSenderTrackException()
      }
    }
  }

  /**
   * Synchronizes the [MediaStreamTrackProxy] of this [RtpSenderProxy] with the underlying
   * [RtpSender].
   */
  private fun syncMediaStreamTrack() {
    val newSenderTrack = obj.track()
    if (newSenderTrack == null) {
      track = null
    } else {
      if (track == null) {
        track = MediaStreamTrackProxy(newSenderTrack)
      } else {
        track!!.replace(newSenderTrack)
      }
    }
  }
}

File: ios/Classes/proxy/RtpSenderProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpSender`, powering it with additional API.
class RtpSenderProxy {
  /// Actual underlying `RTCRtpSender`.
  private var sender: RTCRtpSender

  /// `MediaStreamTrackProxy` of this `RtpSenderProxy`.
  private var track: MediaStreamTrackProxy? = nil

  /// Initializes a new `RtpSenderProxy` for the provided `RTCRtpSender`.
  init(sender: RTCRtpSender) {
    self.sender = sender
    self.syncMediaStreamTrack()
  }

  /// Returns ID of this `RtpSenderProxy`.
  func id() -> String {
    self.sender.senderId
  }

  /// Replaces the `MediaStreamTrackProxy` of the underlying `RtpSender` with
  /// the provided one.
  func replaceTrack(t: MediaStreamTrackProxy?) {
    self.track = t
    self.sender.track = t?.obj()
  }

  /// Synchronizes the `MediaStreamTrackProxy` of this `RtpSenderProxy` with the
  /// underlying `RtpSender`.
  func syncMediaStreamTrack() {
    let newTrack = self.sender.track
    if newTrack == nil {
      self.track = nil
    } else {
      if self.track == nil {
        self.track = MediaStreamTrackProxy(track: newTrack!, deviceId: nil, source: nil)
      }
    }
  }
}
